sharedresource/v1alpha1/types_shared_secret.go

```diff
@@ -0,0 +1,79 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SharedSecret allows a Secret to be shared across namespaces.
+// Pods can mount the shared Secret by adding a CSI volume to the pod specification using the
+// "csi.sharedresource.openshift.io" CSI driver and a reference to the SharedSecret in the volume attributes:
+//
+// spec:
+//  volumes:
+//  - name: shared-secret
+//    csi:
+//      driver: csi.sharedresource.openshift.io
+//      volumeAttributes:
+//        sharedSecret: my-share
+//
+// For the mount to be successful, the pod's service account must be granted permission to 'use' the named SharedSecret object
+// within its namespace with an appropriate Role and RoleBinding. For compactness, here are example `oc` invocations for creating
+// such Role and RoleBinding objects.
+//
+//  `oc create role shared-resource-my-share --verb=use --resource=sharedsecrets.sharedresource.openshift.io --resource-name=my-share`
+//  `oc create rolebinding shared-resource-my-share --role=shared-resource-my-share --serviceaccount=my-namespace:default`
+//
+// Shared resource objects, in this case Secrets, have default permissions of list, get, and watch for system authenticated users.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +kubebuilder:subresource:status
+//
+type SharedSecret struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired shared secret
+	// +kubebuilder:validation:Required
+	Spec SharedSecretSpec `json:"spec,omitempty"`
+
+	// status is the observed status of the shared secret
+	Status SharedSecretStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SharedSecretList contains a list of SharedSecret objects.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type SharedSecretList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []SharedSecret `json:"items"`
+}
+
+// SharedSecretSpec defines the desired state of a SharedSecret
+// +k8s:openapi-gen=true
+type SharedSecretSpec struct {
+	// name represents the name of the Secret that is being referenced.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```

Comment:
This flattening isn't quite what I was looking for. Have a look at https://github.com/kubernetes/community/blob/91bc44bf754992ac65ebfd08a2c3733c42aa1b8e/contributors/devel/sig-architecture/api-conventions.md#naming-conventions where it talks about `FooRef` which seems suitable here with a specific type

```
type ConfigMapReference struct{
  // the help you have for each spec field now.
  Name
  Namespace
}
```

and in SharedConfigMap

```
type SharedConfigMapSpec struct{
    ConfigMapRef ConfigMapReference
}

Reply:
updated in current push, ptal